Padding, effectively providing a silence / counterpoint shunt

twobob opened the pull request:
This should allow for the creation of counterpoint beats and sectional positioning within reason.

PrePad null space before extant clip info, when chained will blank start of result.

carrierdown commented on src/Mutate4l/Commands/PrePad.cs:
```diff
@@ -0,0 +1,53 @@
+    public class PrePadOptions
+    {
+        [OptionInfo(type: OptionType.Default, 1/2f)]
+        public decimal[] Lengths { get; set; } = { 1 };
```
Seems like you're only using the first value here anyway, so just use a standard decimal here instead.

twobob replied:
okay. I /thought/ it maybe future-proofing but fair point. 👍

carrierdown commented on src/Mutate4l/Commands/PrePad.cs:
```diff
@@ -0,0 +1,53 @@
+
+                var processedClip = new Clip(duration, clip.IsLooping);
+
+                System.Collections.Generic.List<NoteEvent> Tempnotes = ClipUtilities.GetSplitNotesInRangeAtPosition(0, duration, clip.Notes, 0);
```
I don't see why this would be needed. In the case of padding a clip, it should be sufficient to iterate through all notes adding the padding-length, and then adding the same length to the total length of the clip. You can discard the tempnotes altogether.

carrierdown commented on src/Mutate4l/Commands/PrePad.cs:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    // # desc: Crops a clip to the desired length, or within the desired region.
+    public static class PrePad
```
I think this command could solve more use cases if it also included a mechanism to set a desired length of the entire clip, where this command would then figure out if it needed to simply crop the clip (i.e. the original clip is too long, or add padding in order to make it the same as the length specified). It could be made into a more general command called Padding, where by default it would add padding as it does currently, but an additional option could make it behave like described above, for instance -fit or something like that.

carrierdown replied:
In the case of cropping, it could simply use the already existing crop command to achieve this.

twobob replied:
I take that on board.  so perhaps it should have the decimal list as arguments then.

Shall digest

twobob commented:
Updated as per your comments.

Renamed to padding, supported second argument, removed unnecessary list, etc 

In the case of the second argument being shorter than the total duration it simply pushes the notes back behind the live clip which is fine and could be handy to create accents on turnarounds.

As you rightly said. You already have a crop.

carrierdown commented:
Thanks, I'll have a closer look at this when I have more time. I might make some small additions/tweaks, hope you don't mind :)

twobob commented:
Not at all, feel free. Am I okay to submit pulls against the current master for other stuff in the interim.  The MidiVelocity (or whatever I called it) could be pushed, as well as the abbreviation ideas.

Monolithic push with those thing in place acceptable? or do you want it breaking up.
Working practices are a mystery to lone wolves.

carrierdown commented:
Preferably break it up into two PR's, but it's not really a big deal, so only split them if it's trivial to do from your side.

carrierdown commented:
Merged to master via commit 2b63b6
